crates/burrego/src/opa/wasm.rs

```diff
@@ -110,36 +118,25 @@ impl Evaluator {
         let memory = Memory::new(&mut store, memory_ty)?;
         linker.define("env", "memory", memory)?;
 
-        // env.opa_abort(addr) void
-        // Called if an internal error occurs.
-        // The addr refers to a null-terminated string in the shared memory buffer.
-        let opa_abort = Func::wrap(
-            &mut store,
-            move |mut caller: Caller<'_, Option<StackHelper>>, addr: i32| {
-                let stack_helper = caller.data().unwrap();
-                let msg = stack_helper
-                    .pull_json(caller.as_context_mut(), &memory, addr)
-                    .unwrap();
-                println!("OPA abort with message: {:?}", msg);
-                std::process::exit(1);
-            },
-        );
-        linker.define("env", "opa_abort", opa_abort)?;
+        macro_rules! opa_callback {
+            ($name: ident) => {
+                let $name = Func::wrap(
+                    &mut store,
+                    move |mut caller: Caller<'_, Option<StackHelper>>, addr: i32| {
+                        let stack_helper = caller.data().unwrap();
+                        let msg = stack_helper
+                            .pull_json(caller.as_context_mut(), &memory, addr)
+                            .unwrap();
+                        (host_callbacks.$name)(msg);
+                    },
+                );
+                linker.define("env", stringify!($name), $name)?;
+            };
+        }
```

Comment:
Wouldn't it be better to move that outside of the `new` function?

Reply:
It's closer to the usage of the macro (only the two next lines actually call it), and the macro captures variables from the lexical context, such as `store` and `memory`. They could be arguments to the macro but the locality of the macro still applies. I can move it if you think is more clear though.

Reply:
Good point, no worries then